### PR TITLE
Add the possibility to set package ensure.

### PR DESCRIPTION
**Pull request by n00by**
The default value is set to installed so the existing behaviour is not
changed.

### CODE REVIEW

**Comment by lbetz** [comment]
What's with the other packages? icinga2-ido-mysql and icinga2-ido-pgsql?

**Comment by Reamer** [comment]
On agent platforms (~99%) you have no icinga2-ido-mysql or icinga2-ido-pgsql package installed. I think we can ignore this.

**Comment by dnsmichi** [comment]
Satellites?

**Comment by n00by** [comment]
Good idea. Will also add the parameters "ido_pgsql_package_ensure" and "ido_mysql_package_ensure"

**Comment by lbetz** [comment]
Please don't do it in the main class, add it to the feature classes.

**Comment by lbetz** [comment]
Wait...
versions must be same. If a new version is installed, the package for ido has to be updated too. But how the the db schema will be updated?

**Comment by Sukiyakijango** [comment]
We do it using an ugly exec(with a subscribe on the package).. like the puppetlabs MySQL module..
The stdout and stderr are redirected to different log files.. plus we gave a check that checks if the error log file is empty.. but having this process working for the last 2 years (other modules).. never had a problem.. 
for icinga we have this process since July.. every thing perfect.. will send the exec on monday.. I use a regen transformation.. to get the version.sql..

**Comment by n00by** [comment]
When I update icinga2 apt seems also to bump the icinga2-ido-pgsql version:

`Commandline: /usr/bin/apt-get -q -y -o DPkg::Options::=--force-confold --force-yes install icinga2=2.6.2-1~ppa1~xenial1
Requested-By: -
Upgrade: icinga2-bin:amd64 (2.6.0-1~ppa1~xenial2, 2.6.2-1~ppa1~xenial1), icinga2-doc:amd64 (2.6.0-1~ppa1~xenial2, 2.6.2-1~ppa1~xenial1), icinga2-ido-pgsql:amd64 (2.6.0-1~ppa1~xenial2, 2.6.2-1~ppa1~xenial1), libicinga2:amd64 (2.6.0-1~ppa1~xenial2, 2.6.2-1~ppa1~xenial1), icinga2-common:amd64 (2.6.0-1~ppa1~xenial2, 2.6.2-1~ppa1~xenial1), icinga2:amd64 (2.6.0-1~ppa1~xenial2, 2.6.2-1~ppa1~xenial1)
`
Not sure how other package tools handle this. Db schema updates need to be done manually atm. Maybe it is possible to add a update_schema to the module that execs "/usr/share/icinga2-ido-pgsql/schema/upgrade/x.x.x.sql" depending on the installed package?

File: manifests/feature/idomysql.pp [hunk 1]
```diff
@@ -7,6 +7,10 @@
 # [*ensure*]
 #   Set to present enables the feature ido-mysql, absent disables it. Defaults to present.
 #
+# [*ido_mysql_package_ensure*]
+#   Specifies the package version to install. Can be 'installed', 'absent',
```
Comment (Reamer) [inline manifests/feature/idomysql.pp]: Maybe the name package_ensure is better, because this parameter is part of class "icinga2::feature::**idomysql**". Same for pgsql

We can also save two parameter definitions in params.pp

Comment (n00by) [reply manifests/feature/idomysql.pp]: True. Updated this.

**Comment by bobapple** [comment]
Related to #244

This is a complex topic and needs to be reviewed carefully. If we cannot come up with a clean implementation I am not willing to add this to the module. I see no point in adding a feature that only works for agents or does not care about schema updates, satellites or whatever.

**Comment by n00by** [comment]
I think its a pretty important and basic feature to manage package versions via puppet. You don't want to bump packages on 500 clients manually. Ofc you need to consider schema updates if you make bigger version jumps on the master, but this is the same if you update manually.

Managing the schema updates could be another enhancement for this module, but isn't necessary sth. linked to this changes.

**Comment by antaflos** [comment]
Being able to manage the package version is absolutely essential and one of the cornerstones of Puppet's very existence. Otherwise this module becomes pretty useless when the next Icinga security or feature update rolls around because there is no way anybody should be forced to upgrade the packages *manually* on hundreds or thousands of Icinga clients (or even servers). Some sites have the luxury of having orchestration tools like MCollective available but that is definitely not something this module should count on.

**Comment by dnsmichi** [comment]
Setting a specific package version inside a Puppet module also requires the availabilty of such packages on repositories. What if you decide to stick with 2.6.2 on your clients and at some point, the package is gone from public repos for e.g. Centos. Adding a new client breaks Puppet, so new installs will fail while the existing packages are installed/cached. You can ofc workaround that issue by building your own package mirror (and you certainly need that when demanding to use versioning and stay safe).

I certainly understand the demand to pin versions. A certain wokaround would be to pin the packages using your packagemanagement tool.

The point here is that it will be required to provide a solution which works for any managed package. "Unfortunately" the IDO feature requires schema imports and/or migrations. It is even hard to manage such manually, as there's no good way of handling smooth upgrades (better example: Icinga Director). 

I'd suggest not to discuss or explain why such a feature is needed (calling something "useless" reads more like frustration than anything helpful), but actually propose a solution to the database schema handling issue. My 2 cents on a possible solution:

- find a way to determine if specific features like ido are enabled for this node
- print a fair warning if a version is specified
- force the version on all involved packages
- watch the reports that the icinga2 service can't be started

It might be worthwhile to consider changing the way how Icinga 2 handles wrong schema versions. Currently it calls exit(), mainly for the reason that users don't read logs and wonder why data is outdated inside web frontends such as Icinga Web 2. More on that in icinga/icinga2#4440 - +1s over there if you agree on it please.

If Icinga 2 would continue running with the specified package version, but the ido feature remains broken (aka no schema migration handling by Puppet), this might help solve the issue. Still I'm thinking that users will open follow-up issues on exactly that new situation and problem - "pinned version breaks ido feature".

**Comment by antaflos** [comment]
One of Puppet's core features is managing installed package versions. This is useful both for making sure a specific package version is (and remains) installed (pinning packages, as you call it) and for upgrading packages to a newer version.

Such packages must obviously be provided by a package management system. Puppet supports quite a few of those, and both Debian and RedHat packages (deb, rpm) and package managers (yum, dpkg, APT) are first-class citizens in this regard.

However, it is neither Puppet's, nor this module's responsibility to manage which packages are available or not available in an upstream package repository. If an operations team requires that a specific package version is always installable then it is their responsibility to manage local repositories that mirror and proxy upstream repositories, just like you say. I wouldn't even call it a workaround; managing artefacts relevant for the infrastructure is (or should be) in the job description of every operations person. We use Artifactory quite successfully for such requirements. 

What I am saying is that this:

> What if you decide to stick with 2.6.2 on your clients and at some point, the package is gone from public repos for e.g. Centos. Adding a new client breaks Puppet, so new installs will fail while the existing packages are installed/cached. 

is not a problem Puppet or this module have to solve, and it is not a valid argument for intentionally disabling the "package ensure" feature Puppet has provided since the beginning. On the contrary, taking this argument further means that older Icinga clients will keep running the Icinga version that was available during installation, while newer Icinga clients (deployed and installed months later) will run a different, newer version of Icinga, and the user has not control over the matter. This is not a great situation for an Icinga site of non-trivial size.

This just to elaborate on my previous comment and why I think being able to manage package versions using a configuration management/automation framework like Puppet (or Chef, Salt, Ansible, etc) does indeed matter.

The database schema handling issue on the other hand is definitely not related to Puppet or this module but seems to me like something that needs to be solved within Icinga itself. I am not very familiar with Icinga's internal architecture or the IDO features so maybe I am misunderstanding some things. But it seems to me that if a database schema upgrade is required then it should be performed automatically, either during Icinga startup or after package installation (e.g. in a `post-install` script). Isn't this a fairly common and by now pretty much a solved problem?

We do similar things in the applications we develop and operate, and I imagine Icinga isn't much different: the application internally knows which database schema version is required; the current schema version is specified in a database table like `schema_info` or similar; an updated application version requires a higher schema version than what the database contains so when the application starts it knows that it must run the database schema upgrade tasks (these tasks should be idempotent -- we use Liquibase for that); if the schema upgrade is successful the application finishes starting; if not the application terminates, writes useful logs and refuses to start up again until the database schema is fixed (in theory this should never happen).

None of that is a concern for configuration management frameworks, and I'd say most CM frameworks can't even handle such issues. I think Puppet should not be used to work around problems that can probably best be solved on the application side.

> Currently it calls exit(), mainly for the reason that users don't read logs and wonder why data is outdated inside web frontends such as Icinga Web 2.

Is that really use case Icinga tries to cover? Users who ignore logs? Wouldn't that be an even more compelling argument for automatically handling schema upgrades?

I suppose the main question for me is this: why doesn't Icinga handle required database schema upgrades automatically? I am sure there was a good reason for it when the decision was made, can you elaborate on that so we can understand better?

**Comment by dnsmichi** [comment]
Some background on the IDO schema:

There's currently no proper way to handle schema updates within the IDO feature. Packages handle that in different ways, i.e. the Debian packages use dbconfig and allow the user to upgrade the schema during package updates. RPM packages do not provide such a mechanism, and neither do other Linux/Unix variants.

So there's some differences with how packages handle such updates themselves. This has been there for many years, with the NDO/IDO schema handling coming from Icinga 1.x. Icinga 2 uses the same schema, with own additions and enhancements (i.e. adding the check source, or endpoint/zone objects). Those are just SQL files providing update statements and bumping the schema version. The version is '1.14.0' atm, which is required by Icinga v2.6.x. If 2.7 requires a change, the schema version will be bumped. 

Then you'll need sort of script or provider which takes care of 

- checking the current database version (icinga_dbversion)
- compare it to the current package version and determine which upgrade files are required being executed
- run the sql update
- ensure everything's working fine

That's basically what dbconfig on Debian does. RPM packages don't do that, and from what I've learned from RPM packagers, it is pretty uncommon to let an RPM update handle database schema upgrades on its own.

If we come up with our own backend and schema, we'll certainly take care making schema migrations easier. History tells that this always is a show stopper and is problematic in many environments, not only with automated deployments.

Imho a Puppet/Ansible/Chef/etc. module should offer a reasonable way to import the schema. It is already there, so the Puppet module allows you to import the current schema. Now you'll have the problem with pinning Icinga 2 to a specific version, and need to handle the schema upgrades. AFAIK the module does not currently provide a way to handle schema upgrades, and adding such will be quite tricky I believe.

The reason I'm jumping in here is to get an idea if there's things needed inside Icinga 2 to make the handling inside the Puppet module easier. Although if we change anything there, this won't help with existing installations (i.e. v2.6.2 being pinned). But it offers a possibility to enhance such things in the future.

TL;DR - if you don't manage the ido's schema version from inside the Puppet module, you can probably safely specify a package version. Otherwise you should be aware of possible upgrade problems.

**Comment by lbetz** [comment]
related #244
